oximeter/db/src/client.rs

```diff
@@ -854,198 +858,524 @@ mod tests {
 
-        recall_measurement_histogram_i8_test(address).await.unwrap();
+        let client = Client::new("127.0.0.1:443".parse().unwrap(), &log);
+        assert!(matches!(
+            client.ping().await,
+            Err(Error::DatabaseUnavailable(_))
+        ));
 
-        recall_measurement_histogram_u8_test(address).await.unwrap();
+        logctx.cleanup_successful();
+        Ok(())
+    }
 
-        recall_measurement_histogram_i16_test(address).await.unwrap();
+    async fn data_is_replicated_test(
+        cluster: &ClickHouseCluster,
+    ) -> Result<(), Error> {
+        let logctx = test_setup_log("test_data_is_replicated");
+        let log = &logctx.log;
 
-        recall_measurement_histogram_u16_test(address).await.unwrap();
+        // Create database in node 1
+        let client_1 = Client::new(cluster.replica_1.address, &log);
+        assert!(client_1.is_oximeter_cluster().await.unwrap());
+        client_1
+            .init_replicated_db()
+            .await
+            .expect("Failed to initialize timeseries database");
 
-        recall_measurement_histogram_i32_test(address).await.unwrap();
+        // Verify database exists in node 2
+        let client_2 = Client::new(cluster.replica_2.address, &log);
+        assert!(client_2.is_oximeter_cluster().await.unwrap());
+        let sql = String::from("SHOW DATABASES FORMAT JSONEachRow;");
+        let result = client_2.execute_with_body(sql).await.unwrap();
 
-        recall_measurement_histogram_u32_test(address).await.unwrap();
+        // Try a few times to make sure data has been synchronised.
+        let tries = 5;
+        for _ in 0..tries {
+            if !result.contains("oximeter") {
+                sleep(Duration::from_secs(1)).await;
+                continue;
+            } else {
+                break;
+            }
+        }
 
-        recall_measurement_histogram_i64_test(address).await.unwrap();
+        assert!(result.contains("oximeter"));
 
-        recall_measurement_histogram_u64_test(address).await.unwrap();
+        // Insert row into one of the tables
+        let sql = String::from(
+            "INSERT INTO oximeter.measurements_string (datum) VALUES ('hiya');",
+        );
+        client_2.execute_with_body(sql).await.unwrap();
 
-        recall_measurement_histogram_f64_test(address).await.unwrap();
+        let sql = String::from(
+            "SELECT * FROM oximeter.measurements_string FORMAT JSONEachRow;",
+        );
+        let result = client_2.execute_with_body(sql.clone()).await.unwrap();
+        assert!(result.contains("hiya"));
 
-        db.cleanup().await.expect("Failed to cleanup ClickHouse server");
+        client_1.wipe_replicated_db().await?;
         logctx.cleanup_successful();
+
+        // TODO(https://github.com/oxidecomputer/omicron/issues/4001): With distributed
+        // engine, it can take a long time to sync the data. This means it's tricky to
+        // test that the data exists on both nodes.
```

Comment:
https://github.com/oxidecomputer/omicron/pull/4149#discussion_r1370775228 didn't really work unfortunately :( I will continue to think about this one, but I don't want to stop this PR as I think it's important to have the rest of the replicated tests running on CI.

Reply:
What do you mean by "didn't work"? The [`FLUSH DISTRIBUTED`](https://clickhouse.com/docs/en/sql-reference/statements/system#flush-distributed) statement makes a call that synchronously flushes all data to the other nodes. IIUC, we could call that here, and if / when it returns, all data should have been replicated.

Reply:
Also, if we're really not going to test that data inserted on one node is visible on the other, we should not be calling this `data_is_replicated_test()`.